localstack/services/s3/provider.py

```diff
@@ -145,6 +171,73 @@ def put_object(
         response: PutObjectOutput = call_moto(context)
         return response
 
+    def add_custom_routes(self):
+        # virtual-host style: https://bucket-name.s3.region-code.amazonaws.com/key-name
+        # host_pattern_vhost_style = f"{bucket}.s3.<regex('({AWS_REGION_REGEX}\.)?'):region>{LOCALHOST_HOSTNAME}:{get_edge_port_http()}"
+        host_pattern_vhost_style = f"<regex('.*'):bucket>.s3.<regex('({AWS_REGION_REGEX}\\.)?'):region>{LOCALHOST_HOSTNAME}<regex('(?::\\d+)?'):port>"
+        ROUTER.add(
+            "/<path:path>",
+            host=host_pattern_vhost_style,
+            endpoint=self.serve_bucket,
+        )
+        ROUTER.add(
+            "/",
+            host=host_pattern_vhost_style,
+            endpoint=self.serve_bucket,
+            defaults={"path": "/"},
+        )
+
+        # regions for path-style need to be parsed correctly
+        host_pattern_vhost_style = f"s3.<regex('({AWS_REGION_REGEX}\\.)'):region>{LOCALHOST_HOSTNAME}<regex('(?::\\d+)?'):port>"
+        ROUTER.add(
+            "/<regex('.+'):bucket>/<path:path>",
+            host=host_pattern_vhost_style,
+            endpoint=self.serve_bucket,
+        )
+        ROUTER.add(
+            "/<regex('.+'):bucket>",
+            host=host_pattern_vhost_style,
+            endpoint=self.serve_bucket,
+            defaults={"path": "/"},
+        )
+
+    def serve_bucket(
+        self, request: Request, bucket: str, path: str, region: str, port: str
+    ) -> Response:
+        # TODO region pattern currently not working -> removing it from url
+        host, rewritten_url = self.rewrite_url(request.url, bucket, region)
+        LOG.debug(
+            f"Rewritten original virtual host url: {request.url} to path-style url: {rewritten_url}"
+        )
+        copied_headers = copy.deepcopy(request.headers)
+        # TODO is_internal_call_context
+        original_host = copied_headers["Host"]
+        copied_headers["Host"] = host
+        params = {"data": request.data, "headers": copied_headers}
+        response = requests.request(method=request.method, url=rewritten_url, **params)
+
+        response.request.headers["Host"] = original_host
+        response.request.url = request.url
+        ls_response = convert_response(response)
+        return ls_response
+
+    def rewrite_url(self, url: str, bucket: str, region: str) -> Tuple[str, str]:
```

Comment:
nit: could we add some pydoc here to explain what the method is used for and what it does?